src/main/java/ch/njol/skript/classes/registry/RegistryUtils.java

```diff
@@ -0,0 +1,129 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.classes.registry;
+
+import ch.njol.skript.localization.Language;
+import ch.njol.skript.localization.Noun;
+import ch.njol.util.NonNullPair;
+import ch.njol.util.StringUtils;
+import org.bukkit.Keyed;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Registry;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+public class RegistryUtils<R extends Keyed> {
+
+	private final Registry<R> registry;
+	private final String languageNode;
+
+	private final Map<R, String> names = new HashMap<>();
+	private final Map<String, R> parseMap = new HashMap<>();
+
+	public RegistryUtils(Registry<R> registry, String languageNode) {
+		assert !languageNode.isEmpty() && !languageNode.endsWith(".") : languageNode;
+		this.registry = registry;
+		this.languageNode = languageNode;
+		refresh();
+		Language.addListener(this::refresh);
+	}
+
+	void refresh() {
+		names.clear();
+		parseMap.clear();
+		for (R registryObject : registry) {
+			NamespacedKey namespacedKey = registryObject.getKey();
+			String namespace = namespacedKey.getNamespace();
+			String key = namespacedKey.getKey();
+			String keyWithSpaces = key.replace("_", " ");
+			String languageKey = languageNode + "." + key;
+
+			// Put the full namespaced key as a pattern
+			parseMap.put(namespacedKey.toString(), registryObject);
+
+			// If the object is a vanilla Minecraft object, we'll add the key with spaces as a pattern
+			if (namespace.equalsIgnoreCase("minecraft")) {
```

Comment:
IIRC there is a field like `NamespacedKey.MINECRAFT` that might be better to compare against.